File: ui/theme.py
"""
App-wide color themes: a light theme (the app's original look) and
several dark variants with different accent-color combos. A `Theme`
drives both the Qt widget stylesheet (QSS) and the matplotlib figure
styling, so plots stay legible and visually consistent with whichever
theme is active instead of staying stuck with light-mode white
backgrounds no matter what the rest of the app looks like.
"""
from __future__ import annotations

from dataclasses import dataclass


@dataclass
class Theme:
    key: str
    label: str
    bg: str            # window/app background
    panel_bg: str       # group box / card background
    text: str
    muted_text: str      # "description"-style secondary labels
    border: str
    input_bg: str
    accent: str           # primary buttons, progress bar, selected tab
    accent_hover: str
    accent_text: str       # text/icon color drawn on top of `accent`
    danger: str             # stop button
    danger_hover: str
    success: str              # paraview button
    success_hover: str
    plot_bg: str               # matplotlib figure/axes background
    plot_fg: str                # matplotlib text/tick/spine color
    plot_grid: str


THEMES: dict[str, Theme] = {
    "light": Theme(
        key="light", label="Light",
        bg="#f3f4f6", panel_bg="#ffffff", text="#1f2937", muted_text="#5a5f6a", border="#d0d3d8", input_bg="#ffffff",
        accent="#2563eb", accent_hover="#1d4ed8", accent_text="#ffffff",
        danger="#dc2626", danger_hover="#b91c1c", success="#059669", success_hover="#047857",
        plot_bg="#ffffff", plot_fg="#1f2937", plot_grid="#e5e7eb",
    ),
    "dark_blue": Theme(
        key="dark_blue", label="Dark - Blue",
        bg="#0f172a", panel_bg="#1e293b", text="#e2e8f0", muted_text="#94a3b8", border="#334155", input_bg="#0f172a",
        accent="#3b82f6", accent_hover="#60a5fa", accent_text="#ffffff",
        danger="#f87171", danger_hover="#ef4444", success="#34d399", success_hover="#10b981",
        plot_bg="#1e293b", plot_fg="#e2e8f0", plot_grid="#334155",
    ),
    "dark_purple": Theme(
        key="dark_purple", label="Dark - Purple",
        bg="#161320", panel_bg="#221d33", text="#e9e4f7", muted_text="#a99fc7", border="#3c3356", input_bg="#161320",
        accent="#a78bfa", accent_hover="#c4b5fd", accent_text="#1e1533",
        danger="#f87171", danger_hover="#ef4444", success="#4ade80", success_hover="#22c55e",
        plot_bg="#221d33", plot_fg="#e9e4f7", plot_grid="#3c3356",
    ),
    "dark_teal": Theme(
        key="dark_teal", label="Dark - Teal",
        bg="#0c1a1a", panel_bg="#132727", text="#dcf5f0", muted_text="#8fbdb5", border="#254545", input_bg="#0c1a1a",
        accent="#2dd4bf", accent_hover="#5eead4", accent_text="#062a26",
        danger="#fb7185", danger_hover="#f43f5e", success="#a3e635", success_hover="#84cc16",
        plot_bg="#132727", plot_fg="#dcf5f0", plot_grid="#254545",
    ),
    "dark_amber": Theme(
        key="dark_amber", label="Dark - Amber",
        bg="#1c1712", panel_bg="#2a221a", text="#f6ecdd", muted_text="#bfa98a", border="#4a3c28", input_bg="#1c1712",
        accent="#fbbf24", accent_hover="#fcd34d", accent_text="#241c0a",
        danger="#f87171", danger_hover="#ef4444", success="#a3e635", success_hover="#84cc16",
        plot_bg="#2a221a", plot_fg="#f6ecdd", plot_grid="#4a3c28",
    ),
}

DEFAULT_THEME_KEY = "dark_blue"


def build_stylesheet(theme: Theme) -> str:
    return f"""
QMainWindow, QWidget {{ background: {theme.bg}; color: {theme.text}; }}
QGroupBox {{
    font-weight: 600;
    border: 1px solid {theme.border};
    border-radius: 8px;
    margin-top: 14px;
    background: {theme.panel_bg};
    padding-top: 6px;
    color: {theme.text};
}}
QGroupBox::title {{
    subcontrol-origin: margin;
    left: 10px;
    padding: 0 4px;
    color: {theme.text};
}}
QLabel {{ color: {theme.text}; background: transparent; }}
QLabel#description {{ color: {theme.muted_text}; font-style: italic; }}
QPushButton {{
    background: {theme.accent};
    color: {theme.accent_text};
    border-radius: 6px;
    padding: 7px 14px;
    font-weight: 600;
    border: none;
}}
QPushButton:hover {{ background: {theme.accent_hover}; }}
QPushButton:disabled {{ background: {theme.border}; color: {theme.muted_text}; }}
QPushButton#stopButton {{ background: {theme.danger}; }}
QPushButton#stopButton:hover {{ background: {theme.danger_hover}; }}
QPushButton#paraviewButton {{ background: {theme.success}; }}
QPushButton#paraviewButton:hover {{ background: {theme.success_hover}; }}
QProgressBar {{
    border: 1px solid {theme.border};
    border-radius: 6px;
    text-align: center;
    background: {theme.input_bg};
    color: {theme.text};
    height: 18px;
}}
QProgressBar::chunk {{ background-color: {theme.accent}; border-radius: 6px; }}
QStatusBar {{ background: {theme.panel_bg}; color: {theme.text}; }}
QLineEdit, QSpinBox, QDoubleSpinBox, QComboBox {{
    background: {theme.input_bg};
    color: {theme.text};
    border: 1px solid {theme.border};
    border-radius: 4px;
    padding: 3px;
    selection-background-color: {theme.accent};
}}
QComboBox QAbstractItemView {{
    background: {theme.input_bg};
    color: {theme.text};
    selection-background-color: {theme.accent};
    selection-color: {theme.accent_text};
}}
QCheckBox, QRadioButton {{ color: {theme.text}; background: transparent; }}
QScrollArea {{ background: transparent; border: none; }}
QTabWidget::pane {{ border: 1px solid {theme.border}; background: {theme.bg}; }}
QTabBar::tab {{
    background: {theme.panel_bg};
    color: {theme.text};
    padding: 8px 16px;
    border: 1px solid {theme.border};
    border-bottom: none;
}}
QTabBar::tab:selected {{ background: {theme.accent}; color: {theme.accent_text}; }}
QMenuBar {{ background: {theme.panel_bg}; color: {theme.text}; }}
QMenuBar::item:selected {{ background: {theme.accent}; color: {theme.accent_text}; }}
QMenu {{ background: {theme.panel_bg}; color: {theme.text}; border: 1px solid {theme.border}; }}
QMenu::item:selected {{ background: {theme.accent}; color: {theme.accent_text}; }}
QMessageBox {{ background: {theme.panel_bg}; color: {theme.text}; }}
QScrollBar:vertical {{ background: {theme.panel_bg}; width: 12px; }}
QScrollBar::handle:vertical {{ background: {theme.border}; border-radius: 5px; min-height: 20px; }}
QScrollBar::handle:vertical:hover {{ background: {theme.accent}; }}
QScrollBar::add-line:vertical, QScrollBar::sub-line:vertical {{ height: 0px; }}
"""


def apply_theme_to_axes(ax, theme: Theme) -> None:
    """Re-applies theme coloring to a matplotlib Axes -- needed after
    every `.cla()`, since clearing an axes resets it to matplotlib's
    default (light) styling regardless of the figure's own facecolor."""
    ax.set_facecolor(theme.plot_bg)
    ax.title.set_color(theme.plot_fg)
    ax.xaxis.label.set_color(theme.plot_fg)
    ax.yaxis.label.set_color(theme.plot_fg)
    ax.tick_params(colors=theme.plot_fg, which="both")
    for spine in ax.spines.values():
        spine.set_color(theme.plot_grid)


def apply_theme_to_figure(figure, theme: Theme) -> None:
    figure.patch.set_facecolor(theme.plot_bg)
    for ax in figure.get_axes():
        apply_theme_to_axes(ax, theme)

File: cfd_studio_cpp/app/src/theme.cpp
#include "theme.hpp"

const std::array<Theme, 5>& themes() {
    static const std::array<Theme, 5> all = {{
        {
            "light", "Light",
            "#f3f4f6", "#ffffff", "#1f2937", "#5a5f6a", "#d0d3d8", "#ffffff",
            "#2563eb", "#1d4ed8", "#ffffff",
            "#dc2626", "#b91c1c", "#059669", "#047857",
            "#ffffff", "#1f2937", "#e5e7eb",
        },
        {
            "dark_blue", "Dark - Blue",
            "#0f172a", "#1e293b", "#e2e8f0", "#94a3b8", "#334155", "#0f172a",
            "#3b82f6", "#60a5fa", "#ffffff",
            "#f87171", "#ef4444", "#34d399", "#10b981",
            "#1e293b", "#e2e8f0", "#334155",
        },
        {
            "dark_purple", "Dark - Purple",
            "#161320", "#221d33", "#e9e4f7", "#a99fc7", "#3c3356", "#161320",
            "#a78bfa", "#c4b5fd", "#1e1533",
            "#f87171", "#ef4444", "#4ade80", "#22c55e",
            "#221d33", "#e9e4f7", "#3c3356",
        },
        {
            "dark_teal", "Dark - Teal",
            "#0c1a1a", "#132727", "#dcf5f0", "#8fbdb5", "#254545", "#0c1a1a",
            "#2dd4bf", "#5eead4", "#062a26",
            "#fb7185", "#f43f5e", "#a3e635", "#84cc16",
            "#132727", "#dcf5f0", "#254545",
        },
        {
            "dark_amber", "Dark - Amber",
            "#1c1712", "#2a221a", "#f6ecdd", "#bfa98a", "#4a3c28", "#1c1712",
            "#fbbf24", "#fcd34d", "#241c0a",
            "#f87171", "#ef4444", "#a3e635", "#84cc16",
            "#2a221a", "#f6ecdd", "#4a3c28",
        },
    }};
    return all;
}

const Theme& theme_by_key(const QString& key) {
    for (const auto& t : themes()) {
        if (t.key == key) return t;
    }
    for (const auto& t : themes()) {
        if (t.key == QString(kDefaultThemeKey)) return t;
    }
    return themes().front();
}

QString ui_scale_key(UiScale scale) {
    switch (scale) {
        case UiScale::Small: return "small";
        case UiScale::Large: return "large";
        case UiScale::Medium: default: return "medium";
    }
}

QString ui_scale_label(UiScale scale) {
    switch (scale) {
        case UiScale::Small: return "Small";
        case UiScale::Large: return "Large";
        case UiScale::Medium: default: return "Medium";
    }
}

UiScale ui_scale_by_key(const QString& key) {
    if (key == "small") return UiScale::Small;
    if (key == "large") return UiScale::Large;
    return kDefaultUiScale;
}

const std::array<UiScale, 3>& ui_scales() {
    static const std::array<UiScale, 3> all = {UiScale::Small, UiScale::Medium, UiScale::Large};
    return all;
}

namespace {
// Font size / padding metrics per UiScale -- kept as plain string literals
// (not computed) so each tier's numbers are easy to eyeball and tweak.
struct ScaleMetrics {
    const char* base_font_pt;
    const char* button_padding;
    const char* input_padding;
    const char* tab_padding;
};

ScaleMetrics scale_metrics(UiScale scale) {
    switch (scale) {
        case UiScale::Small: return {"8.5pt", "5px 10px", "2px", "5px 10px"};
        case UiScale::Large: return {"14pt", "14px 26px", "7px", "14px 24px"};
        case UiScale::Medium:
        default: return {"10.5pt", "7px 14px", "3px", "8px 16px"};
    }
}
} // namespace

QString build_stylesheet(const Theme& t, UiScale scale, bool dyslexia_font) {
    ScaleMetrics m = scale_metrics(scale);

    QString fontFamilyRule;
    if (dyslexia_font) {
        // Neither font is bundled with the app: OpenDyslexic only takes
        // effect if the user has separately installed it (a specialized
        // dyslexia-readability font, so not something to silently force on
        // everyone via a bundled resource); Comic Sans MS ships with
        // Windows and is a commonly-cited free fallback with the same
        // "distinguishable letterforms" property. Falls through to the
        // platform default font if neither is present.
        fontFamilyRule = "QWidget { font-family: \"OpenDyslexic\", \"Comic Sans MS\", sans-serif; }\n";
    }

    return (fontFamilyRule + QString(
               "QWidget { font-size: %14; }\n"
               "QMainWindow, QWidget { background: %1; color: %2; }\n"
               "QGroupBox {\n"
               "    font-weight: 600;\n"
               "    border: 1px solid %3;\n"
               "    border-radius: 8px;\n"
               "    margin-top: 14px;\n"
               "    background: %4;\n"
               "    padding-top: 6px;\n"
               "    color: %2;\n"
               "}\n"
               "QGroupBox::title {\n"
               "    subcontrol-origin: margin;\n"
               "    left: 10px;\n"
               "    padding: 0 4px;\n"
               "    color: %2;\n"
               "}\n"
               "QLabel { color: %2; background: transparent; }\n"
               "QLabel#description { color: %5; font-style: italic; }\n"
               "QPushButton {\n"
               "    background: %6;\n"
               "    color: %7;\n"
               "    border-radius: 6px;\n"
               "    padding: %15;\n"
               "    font-weight: 600;\n"
               "    border: none;\n"
               "}\n"
               "QPushButton:hover { background: %8; }\n"
               "QPushButton:disabled { background: %3; color: %5; }\n"
               "QPushButton#stopButton { background: %9; }\n"
               "QPushButton#stopButton:hover { background: %10; }\n"
               "QPushButton#paraviewButton { background: %11; }\n"
               "QPushButton#paraviewButton:hover { background: %12; }\n"
               // Without these, the #id background rules above (higher
               // specificity than the plain ":disabled" rule) kept the
               // full-saturation red/green background on a disabled
               // stop/paraview button while still picking up ":disabled"'s
               // muted-gray text color for want of any more-specific color
               // rule -- muted gray on saturated green/red read as low
               // contrast, close to illegible. Falling back to the same
               // neutral disabled look as every other button fixes that
               // and also reads as "disabled" more clearly than a
               // still-vivid button did.
               "QPushButton#stopButton:disabled, QPushButton#paraviewButton:disabled {\n"
               "    background: %3;\n"
               "    color: %5;\n"
               "}\n"
               "QProgressBar {\n"
               "    border: 1px solid %3;\n"
               "    border-radius: 6px;\n"
               "    text-align: center;\n"
               "    background: %13;\n"
               "    color: %2;\n"
               "    height: 18px;\n"
               "}\n"
               "QProgressBar::chunk { background-color: %6; border-radius: 6px; }\n"
               "QStatusBar { background: %4; color: %2; }\n"
               "QLineEdit, QSpinBox, QDoubleSpinBox, QComboBox {\n"
               "    background: %13;\n"
               "    color: %2;\n"
               "    border: 1px solid %3;\n"
               "    border-radius: 4px;\n"
               "    padding: %16;\n"
               "    selection-background-color: %6;\n"
               "}\n"
               "QComboBox QAbstractItemView {\n"
               "    background: %13;\n"
               "    color: %2;\n"
               "    selection-background-color: %6;\n"
               "    selection-color: %7;\n"
               "}\n"
               "QCheckBox, QRadioButton { color: %2; background: transparent; }\n"
               "QScrollArea { background: transparent; border: none; }\n"
               "QTabWidget::pane { border: 1px solid %3; background: %1; }\n"
               "QTabBar::tab {\n"
               "    background: %4;\n"
               "    color: %2;\n"
               "    padding: %17;\n"
               "    border: 1px solid %3;\n"
               "    border-bottom: none;\n"
               "}\n"
               "QTabBar::tab:selected { background: %6; color: %7; }\n"
               "QMenuBar { background: %4; color: %2; }\n"
               "QMenuBar::item:selected { background: %6; color: %7; }\n"
               "QMenu { background: %4; color: %2; border: 1px solid %3; }\n"
               "QMenu::item:selected { background: %6; color: %7; }\n"
               "QMessageBox { background: %4; color: %2; }\n"
               "QScrollBar:vertical { background: %4; width: 12px; }\n"
               "QScrollBar::handle:vertical { background: %3; border-radius: 5px; min-height: 20px; }\n"
               "QScrollBar::handle:vertical:hover { background: %6; }\n"
               "QScrollBar::add-line:vertical, QScrollBar::sub-line:vertical { height: 0px; }\n")
        .arg(t.bg)          // %1
        .arg(t.text)        // %2
        .arg(t.border)      // %3
        .arg(t.panel_bg)    // %4
        .arg(t.muted_text)  // %5
        .arg(t.accent)      // %6
        .arg(t.accent_text) // %7
        .arg(t.accent_hover)// %8
        .arg(t.danger)      // %9
        .arg(t.danger_hover)// %10
        .arg(t.success)     // %11
        .arg(t.success_hover) // %12
        .arg(t.input_bg)    // %13
        .arg(m.base_font_pt)    // %14
        .arg(m.button_padding)  // %15
        .arg(m.input_padding)   // %16
        .arg(m.tab_padding));   // %17
}
